### ISSUE TITLE
Are W-types definable?

### ISSUE THREAD

**Issue by DanGrayson** [UniMath/UniMath#983]
A comment from Thorsten Altenkirch:

In Moerdijk and Gambino’s “Wellfounded trees in categories” it is shown that very topos has W-types. I think that UniMath gives you a topos (on sets) so at least you should get W-sets. Whether you also get W-types is an interesting question.

**Comment by benediktahrens** [comment]
On 31/05/18 15:54, Daniel R. Grayson wrote:
> A comment from Thorsten Altenkirch:
> 
> In Moerdijk and Gambino’s “Wellfounded trees in categories” it is shown 
> that very topos has W-types. I think that UniMath gives you a topos (on 
> sets) so at least you should get W-sets. Whether you also get W-types is 
> an interesting question.

Yes, W-types can be constructed, see 
https://www.ps.uni-saarland.de/~rech/bachelor/thesis.pdf .

**Comment by DanGrayson** [comment]
It probably would be good to implement that, eventually.

**Comment by mortberg** [comment]
For W-sets I think many of the results are already present in https://github.com/UniMath/UniMath/tree/master/UniMath/CategoryTheory/Chains. With this we can do many concrete inductive types (see https://github.com/UniMath/UniMath/tree/master/UniMath/CategoryTheory/Inductives), but I didn't think too hard about using it for defining W-sets. Defining W-types as in the thesis that Benedikt linked to above seems like the best approach to me.

I think @siddharthist was working on this kind of stuff for his Master's thesis?

**Comment by siddharthist** [comment]
I agree, I think @mortberg's link contains the relevant results for W-sets. 

My (undergraduate) thesis formalized some of Benedikt et al's paper "Non-wellfounded trees", which gives the existence of all M-types, not just sets.

**Comment by mortberg** [comment]
@siddharthist : Nice! Have all of your results been merged into UniMath? I haven't really been keeping track of UniMath developments lately. If they are all there I don't think there is too much left to do to get W-types as in Rech's thesis: we just have to define a predicate `isWellFounded` and carve out those M-types that satisfy this predicate.

**Comment by siddharthist** [comment]
@mortberg All of my results have been merged, but I wasn't able to completely formalize Benedikt et al's main proof (just all the lemmas). Details on my proofs and progress are in my thesis, which I'm happy to email along.

**Comment by mortberg** [comment]
Ok, thanks for the update! Was there some deep mathematical problem or just lack of time/stamina for not finishing the formalization?

**Comment by siddharthist** [comment]
Lack of time + stamina!
